File: jarvis/ui/native_drop.py
"""Real paths for files and folders dropped ONTO the desktop WebView.

A web page never learns where a dropped folder lives: the browser hands it a
``File`` with a name and nothing else, by design. In a plain browser tab that is
the end of it and the UI falls back to searching for the name. Inside the
desktop shell we can do better, because the host process sees the native drop
and knows the full path — pywebview exposes it as ``pywebviewFullPath`` on every
file of a ``drop`` event delivered to a Python handler, on Windows (WebView2),
macOS (WKWebView) and Linux (WebKitGTK) alike.

So this module registers exactly such a handler on each window's ``window``
object and, whenever it fires, tells the page: it dispatches a
``jarvis-native-drop`` DOM event whose ``detail`` carries the resolved paths
(and the names they belong to). Any component that handles drops can wait a
moment for that event and use the exact path instead of guessing — the folder
picker in the Agentic IDE does. Outside the shell the event never comes and the
component simply keeps its browser fallback.

Registered on ``window``, not ``body``: pywebview serialises the whole event
including ``currentTarget``, and for ``window`` that is one short string where
``body`` would be the entire document.
"""

from __future__ import annotations

import json
import logging
from collections.abc import Mapping
from typing import Any

log = logging.getLogger(__name__)

# The DOM event the page listens for. Kept in sync with the frontend helper
# (src/lib/nativeDrop.ts).
EVENT_NAME = "jarvis-native-drop"


def dropped_paths(event: Mapping[str, Any]) -> tuple[list[str], list[str]]:
    """The full paths and names of the files in a serialised ``drop`` event.

    Pure, so it can be tested without a window. Files pywebview could not
    resolve (no ``pywebviewFullPath``) are skipped rather than reported as an
    empty path — an empty path is not a hint the page can act on.
    """
    data_transfer = event.get("dataTransfer")
    files = data_transfer.get("files") if isinstance(data_transfer, Mapping) else None
    paths: list[str] = []
    names: list[str] = []
    for item in files or []:
        if not isinstance(item, Mapping):
            continue
        path = item.get("pywebviewFullPath")
        if not isinstance(path, str) or not path:
            continue
        paths.append(path)
        name = item.get("name")
        names.append(name if isinstance(name, str) else "")
    return paths, names


def announce_script(paths: list[str], names: list[str]) -> str:
    """The JavaScript that tells the page what was dropped."""
    detail = json.dumps({"paths": paths, "names": names})
    return f"window.dispatchEvent(new CustomEvent({json.dumps(EVENT_NAME)}, {{detail: {detail}}}));"


def register_native_drop(window: Any) -> bool:
    """Wire the drop handler into ``window``'s CURRENT page.

    Call from the window's ``loaded`` hook — pywebview forgets DOM handlers on
    every load, so this must run again each time the page (re)loads. Never
    raises: a shell without the DOM API (or a stub window in tests) just does
    not get the feature, and the browser fallback still stands.
    """
    dom = getattr(window, "dom", None)
    if dom is None:
        log.debug("native_drop: this window has no DOM bridge — browser fallback stays")
        return False

    def _on_drop(event: Mapping[str, Any]) -> None:
        try:
            paths, names = dropped_paths(event)
            if not paths:
                return
            window.evaluate_js(announce_script(paths, names))
        except Exception:  # noqa: BLE001 — a drop must never take the shell down
            log.exception("native_drop: could not hand the dropped paths to the page")

    try:
        dom.window.on("drop", _on_drop)
    except Exception as exc:  # noqa: BLE001 — pywebview internals vary by backend
        log.warning("native_drop: drop bridge unavailable in this window (%s)", exc)
        return False
    log.debug("native_drop: drop bridge registered")
    return True


__all__ = ["EVENT_NAME", "announce_script", "dropped_paths", "register_native_drop"]

File: jarvis/ui/web/frontend/src/lib/nativeDrop.ts
/**
 * Real paths for files and folders dropped onto the desktop shell.
 *
 * A browser hands a page only the NAME of a dropped folder — never its path.
 * Inside the desktop shell the host process sees the native drop and knows the
 * path; `jarvis/ui/native_drop.py` dispatches it to the page as a
 * `jarvis-native-drop` event right after the DOM `drop`. A drop handler that
 * wants the exact path calls `waitForNativeDrop()` synchronously from its drop
 * handler and gets the paths a moment later — or `null`, promptly, in a plain
 * browser tab or when the shell could not resolve them, and then keeps its own
 * fallback (searching by name).
 */

/** Event name — MUST match `EVENT_NAME` in jarvis/ui/native_drop.py. */
export const NATIVE_DROP_EVENT = "jarvis-native-drop";

/** How long a page waits for the shell before falling back. The round trip is
 * usually well under 200 ms; the ceiling only matters when the bridge is not
 * there at all. */
const NATIVE_DROP_TIMEOUT_MS = 2000;

export interface NativeDropDetail {
  /** Full native paths, one per dropped file or folder that could be resolved. */
  paths: string[];
  /** The browser-side names the paths belong to, in the same order. */
  names: string[];
}

/** True inside the pywebview desktop shell (any OS), where the host can answer. */
export function inDesktopShell(): boolean {
  const host = window as unknown as {
    __JARVIS_EMBEDDED_DESKTOP?: boolean;
    pywebview?: unknown;
  };
  return Boolean(host.__JARVIS_EMBEDDED_DESKTOP || host.pywebview);
}

/**
 * Wait for the shell to report the paths of the drop that just happened.
 *
 * Call it synchronously inside the `drop` handler, so the listener is in place
 * before the shell answers. Resolves `null` immediately outside the shell, and
 * after the timeout when no matching announcement arrives. When `name` is
 * given, an announcement is accepted only if it carries that name — a stray
 * event from another drop is never mistaken for this one.
 */
export function waitForNativeDrop(opts: {
  name?: string;
  timeoutMs?: number;
} = {}): Promise<NativeDropDetail | null> {
  if (!inDesktopShell()) return Promise.resolve(null);
  const timeoutMs = opts.timeoutMs ?? NATIVE_DROP_TIMEOUT_MS;
  return new Promise((resolve) => {
    let timer: number | undefined;
    const onEvent = (event: Event) => {
      const detail = (event as CustomEvent<Partial<NativeDropDetail>>).detail;
      const paths = Array.isArray(detail?.paths) ? detail.paths : [];
      const names = Array.isArray(detail?.names) ? detail.names : [];
      if (paths.length === 0) return;
      if (opts.name && names.length > 0 && !names.includes(opts.name)) return;
      window.removeEventListener(NATIVE_DROP_EVENT, onEvent);
      if (timer !== undefined) window.clearTimeout(timer);
      resolve({ paths, names });
    };
    window.addEventListener(NATIVE_DROP_EVENT, onEvent);
    timer = window.setTimeout(() => {
      window.removeEventListener(NATIVE_DROP_EVENT, onEvent);
      resolve(null);
    }, timeoutMs);
  });
}
